src/npc/src/dialogue.c

```diff
@@ -0,0 +1,221 @@
+#include "npc/dialogue.h"
+
+// STRUCT FUNCTIONS -----------------------------------------------------------
+/* See dialogue.h */
+int convo_init(convo_t *c, char *farewell)
+{
+    assert(c != NULL);
+    strcpy(c->farewell, farewell);
+    c->node_count = 0;
+    c->nodes = NULL;
+
+    return SUCCESS;
+}
+
+/* See dialogue.h */
+convo_t *convo_new(char *farewell)
+{
+    convo_t *c;
+    c = (convo_t*)malloc(sizeof(convo_t));
+    memset(c, 0, sizeof(convo_t));
+    c->farewell = (char*)malloc(MAX_FW_LEN);
+
+    int check = convo_init(c, farewell);
+
+    if (c == NULL || c->farewell == NULL || check != SUCCESS)
+    {
+        return NULL;
+    }
+
+    return c;
+}
+
+/* See LINKED LIST FUNCTIONS section of dialogue.h */
+int delete_all_nodes(node_list_t *nodes)
+{
+    node_list_t *elt, *tmp;
+    DL_FOREACH_SAFE(nodes, elt, tmp)
+    {
+        DL_DELETE(nodes, elt);
+        free(elt);
+    }
+    return SUCCESS;
+}
+
+/* See dialogue.h */
+int convo_free(convo_t *c)
+{
+    free(c->farewell);
+    delete_all_nodes(c->nodes);
+    free(c);
+    return SUCCESS;
+}
+
+/* See dialogue.h */
+int node_init(node_t *n, char *node_id, char *dialogue)
+{
+    assert(n != NULL);
+    strncpy(n->node_id, node_id, strlen(node_id));
+    strncpy(n->dialogue, dialogue, strlen(dialogue));
+    n->connection_count = 0;
+    n->edges = NULL;
+
+    return SUCCESS;
+}
+
+/* See dialogue.h */
+node_t *node_new(char *node_id, char *dialogue)
+{
+    node_t *n = (node_t*)malloc(sizeof(node_t));
+    memset(n, 0, sizeof(node_t));
+    n->node_id = (char*)malloc(MAX_ID_LEN);
+    n->dialogue = (char*)malloc(MAX_DIA_LEN);
+
+    int check = node_init(n, node_id, dialogue);
+    
+    if (n == NULL || n->node_id == NULL || 
+        n->dialogue == NULL || check != SUCCESS)
+    {
+        return NULL;
+    }
+
+    return n;
+}
+
+/* See LINKED LIST FUNCTIONS section of dialogue.h */
+int delete_all_edges(edge_hash_t *edges)
+{
+    edge_t *cur_edge, *tmp;
+    HASH_ITER(hh, edges, cur_edge, tmp)
+    {
+        HASH_DEL(edges, cur_edge);
+        edge_free(cur_edge);
+    }
+    return SUCCESS;
+}
+
+/* See dialogue.h */
+int node_free(node_t *n)
+{
+    assert(n != NULL);
+
+    free(n->node_id);
+    free(n->dialogue);
+    delete_all_edges(n->edges);
+    free(n);
+
+    return SUCCESS;
+}
+
+/* See dialogue.h */
+int edge_init(edge_t *e, node_t *toward, char *keyword, char *quip)
+{
+    assert(e != NULL);
+    strncpy(e->keyword, keyword, strlen(keyword));
+    strncpy(e->quip, quip, strlen(quip));
+    e->toward = toward;
+
+    return SUCCESS;
+}
+
+/* See dialogue.h */
+edge_t *edge_new(node_t *toward, char *keyword, char *quip)
+{
+    edge_t *e = (edge_t*)malloc(sizeof(edge_t));
+    memset(e, 0, sizeof(edge_t));
+    e->keyword = (char*)malloc(MAX_KEY_LEN);
+    e->quip = (char*)malloc(MAX_QUIP_LEN);
```

Comment:
The values within these need to be initialized (to NULL characters); otherwise, this results in potential memory issues. Per the valgrind tests, these would be uninitialized values created by heap allocations and conditional jumps or moves that depend on uninitialized values. I suggest we use `calloc()` to remedy any potential problems. 
```suggestion
    e->keyword = calloc(1, MAX_KEY_LEN);
    e->quip = calloc(1, MAX_QUIP_LEN);
```